### PR TITLE
Add max_identifier_length read-only session setting

### PR DESCRIPTION
**Pull request by hlcianfagna**
## Summary of the changes / Why this improves CrateDB

Closes: #14743

## Checklist

 - [X] Added an entry in `CHANGES.txt` for user facing changes
 - [X] Updated documentation & `sql_features` table for user facing changes
 - [X] Touched code is covered by tests
 - [X] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [X] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by matriv** [comment]
thx @hlcianfagna!
I'm not sure about the value `63` though, as for PostgreSQL indeed it's the maximum identifier length, but that's not the case for CrateDB:
psql:
```
matriv=> create table asdfghjkqwertyuiasdfghjkqwertyuiqwertyuiqwertyuiqwertyuiqwertyu(a int);
CREATE TABLE
matriv=> create table asdfghjkqwertyuiasdfghjkqwertyuiqwertyuiqwertyuiqwertyuiqwertyuaaa(a int);
NOTICE:  identifier "asdfghjkqwertyuiasdfghjkqwertyuiqwertyuiqwertyuiqwertyuiqwertyuaaa" will be truncated to "asdfghjkqwertyuiasdfghjkqwertyuiqwertyuiqwertyuiqwertyuiqwertyu"
ERROR:  relation "asdfghjkqwertyuiasdfghjkqwertyuiqwertyuiqwertyuiqwertyuiqwertyu" already exists
```
CrateDB:
cr> create table asdfghjkqwertyuiasdfghjkqwertyuiqwertyuiqwertyuiqwertyuiqwertyuaaa_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa_aaaaaaaaaaaaaaa(aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa_bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb_cccc
    ccccccccccccccccccccccccccccccccc_dddddddddddddddddddddddddddddddddddddddd_eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_ffffffffffffffffffffffffffffff int);
CREATE OK, 1 row affected (0.060 sec)
cr> select aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa_bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb_ccccccccccccccccccccccccccccccccccccc_dddddddddddddddddddddddddddddddddddddddd_eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_ffffffffffffffffffffffffffffff
    from asdfghjkqwertyuiasdfghjkqwertyuiqwertyuiqwertyuiqwertyuiqwertyuaaa_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa_aaaaaaaaaaaaaaa;
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa_bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb_ccccccccccccccccccccccccccccccccccccc_dddddddddddddddddddddddddddddddddddddddd_eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee_ffffffffffffffffffffffffffffff |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
SELECT 0 rows in set (0.005 sec)
```

**Comment by hlcianfagna** [comment]
> I'm not sure about the value 63
You are right @matriv , @proddata also pointed this out and suggested the value may be around 138 for CrateDB but we were not 100% sure, could you advise what the correct value could be for CrateDB?

**Comment by matriv** [comment]
Normally we allow the java string's max length, but I don't know how a bigger value there can affect postgres tools. We could safely go to `256` which I've tested to verify and it works, and it's more than enough for users to define table/column/index names.

@mfussenegger what do you think?

**Comment by hlcianfagna** [comment]
Going somewhere around 255 I get an InvalidRelationName but it does not seem to be a fix value?

**Comment by matriv** [comment]
Yep, true, apologies, I've tested with `255` missed that last char in copy paste. Indeed that's our limit then: https://github.com/crate/crate/blob/master/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java#L108

File: docs/config/session.rst [hunk 1]
```diff
@@ -164,6 +164,19 @@ releases.
      The session setting has no effect in CrateDB and exists for compatibility
      with ``PostgreSQL``.
 
+.. _conf-session-max_identifier_length:
+
+**max_identifier_length**
+  | *Default:* ``63``
+  | *Modifiable:* ``no``
+
+  Shows the maximum length of object identifiers in bytes.
```
Comment (seut) [inline docs/config/session.rst:173]: This is about all identifiers, not sure what is meant by `object identifiers`.
```suggestion
  Shows the maximum length of identifiers in bytes.
```

**Comment by matriv** [comment]
@seut Do you agree with setting this to `255`, instead of posgres' `63`? see: https://github.com/crate/crate/pull/14745#issuecomment-1735075789

**Comment by seut** [comment]
> @seut Do you agree with setting this to `255`, instead of posgres' `63`? see: [#14745 (comment)](https://github.com/crate/crate/pull/14745#issuecomment-1735075789)

Yes, otherwise it would be misleading as it would not be the true limit ;)

File: server/src/main/java/io/crate/metadata/settings/session/SessionSettingRegistry.java [hunk 2]
```diff
@@ -159,6 +160,18 @@ public SessionSettingRegistry(Set<SessionSettingProvider> sessionSettingProvider
...
                      DataTypes.INTEGER))
+            .put(MAX_IDENTIFIER_LENGTH,
+                 new SessionSetting<>(
+                     MAX_IDENTIFIER_LENGTH,
+                     objects -> {},
+                     Function.identity(),
+                     (s, v) -> {
+                         throw new UnsupportedOperationException("\"" + MAX_IDENTIFIER_LENGTH + "\" cannot be changed.");
+                     },
+                     s -> String.valueOf(63),
```
Comment (matriv) [inline server/src/main/java/io/crate/metadata/settings/session/SessionSettingRegistry.java:171]: Please use directly the constant: https://github.com/crate/crate/blob/master/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java#L108
(also below)

File: docs/config/session.rst [hunk 3]
```diff
@@ -164,6 +164,19 @@ releases.
...
+
+**max_identifier_length**
+  | *Default:* ``63``
+  | *Modifiable:* ``no``
+
+  Shows the maximum length of identifiers in bytes.
+
+  .. NOTE::
+
+     The session setting has no effect in CrateDB and exists for compatibility
```
Comment (hlcianfagna) [inline docs/config/session.rst:177]: Should we remove this note as we move to use `MAX_INDEX_NAME_BYTES` @seut @matriv ?

**Review by matriv — APPROVED** [review]
Thank you @hlcianfagna 

Please adjust the commit message using the header of the PR and the reference to the issue it closes (`Closes: #14743`).

**Pull request merged**